sql/core/src/test/scala/org/apache/spark/sql/DataFrameStatSuite.scala

```diff
@@ -366,6 +366,27 @@ class DataFrameStatSuite extends QueryTest with SharedSparkSession {
     }
   }
 
+  test("SPARK-28818: Respect original column nullability in `freqItems`") {
+    val rows = spark.sparkContext.parallelize(
+      Seq(Row("1", "a"), Row("2", null), Row("3", "b"))
+    )
+    val schema = StructType(Seq(
+      StructField("non_null", StringType, false),
+      StructField("nullable", StringType, true)
+    ))
+    val df = spark.createDataFrame(rows, schema)
+
+    val result = df.stat.freqItems(df.columns)
+
+    val nonNullableFreqItems = result.schema("non_null_freqItems").dataType.asInstanceOf[ArrayType]
+    val nullableFreqItems = result.schema("nullable_freqItems").dataType.asInstanceOf[ArrayType]
+
+    assert(nonNullableFreqItems.containsNull == false)
+    assert(nullableFreqItems.containsNull == true)
+    // Original bug was a NullPointerException exception caused by calling collect(), test for this
+    val resultRows = result.collect()
```

Comment:
Can we assert the results as well?

Reply:
I can if people want me to? My general rule is only to assert exactly what the unit test is trying to test. In this case a specific bug. If unit tests are overlapping it can be very hard to tell what exactly is failing and where a regression has been introduced because all test fail at once.

Reply:
Nullability might have been intentionally disabled on purpose. If the results are incorrect, there's no point of fixing this NPE.

If you want to keep this to test NPE, can we add another test to assert if the results are correct?

Reply:
That's a reasonable argument and I think it warrants being in this test. I'll add it! Thanks 😄